### PR TITLE
Fix torch ONNX export for dynamic input shapes

### PR DESCRIPTION
**Pull request by MarcosAsh**
## Description

On the torch backend, `export_onnx` builds `dynamic_axes` for every `None` dimension (added in #22086, refined in #22737), but the docstring still says dynamic shapes are unsupported and the tests are skipped on torch. Two problems sat underneath:

1. **Recurrent layers silently break with a dynamic time axis.** TorchScript bakes the build-time sequence length into the graph, so an RNN exported with a dynamic time dimension returns wrong outputs for other lengths or crashes at runtime. Non-recurrent models export fine, and a dynamic batch dimension alone is always fine.

2. **The warning suppression never applied to the export.** The `with warnings.catch_warnings()` block wrapped only the `filterwarnings()` setup and exited before `torch.onnx.export` ran, so the `TracerWarning` filters were gone by export time. Under a strict warning filter this turns a routine tracer warning into an error mid-trace, which is why the tests were skipped on torch.

This PR raises a clear `ValueError` when the model has a recurrent layer (`isinstance(layer, RNN)` over `model._flatten_layers()`, so `Bidirectional` and stacked RNNs are covered) and a dynamic non-batch axis, moves the export inside the `catch_warnings` block so the filters take effect, and updates the docstring.

Measured on torch, `onnxruntime` vs eager across lengths:

| model | dynamic axis | behavior on master |
|---|---|---|
| `Conv2D` | batch / height / width | correct |
| `Dense` | sequence | correct |
| `LSTM` | sequence | wrong output (max abs diff 0.76) |
| `GRU` | sequence | wrong output (max abs diff 1.15) |
| `SimpleRNN` | sequence | runtime crash (Split node) |

After this PR the correct rows are unchanged and the broken rows raise at export time.

### Tests

Drops the torch skips on `test_dynamic_shapes_export` and `test_multi_input_dynamic_shapes` (they pass once warnings are suppressed), and adds `test_dynamic_sequence_recurrent_raises` and `test_static_sequence_recurrent_dynamic_batch_exports`. Full `onnx_test.py` passes on torch and jax.

## Contributor Agreement

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

### CODE REVIEW

File: keras/src/export/onnx.py [hunk 1]
```diff
@@ -144,6 +149,35 @@ def export_onnx(
...
+            if recurrent_layers:
+                raise ValueError(
+                    "ONNX export with the torch backend does not support a "
+                    "dynamic time dimension for recurrent layers (found: "
+                    f"{recurrent_layers}). TorchScript bakes in the sequence "
+                    "length at export time, which yields incorrect results "
+                    "for other lengths. Provide a static sequence length in "
+                    "`input_signature`; only the batch dimension may be "
+                    "dynamic for models with recurrent layers."
+                )
```
[the remark this answers is not in the document]
Comment (MarcosAsh) [reply keras/src/export/onnx.py]: Good catch, fixed. The guard now inspects each recurrent layer's own sequence shape and only raises when that layer's time axis is dynamic, so a static-sequence RNN sitting next to a dynamic non-recurrent input (e.g. a `Conv2D` with dynamic spatial dims) no longer trips it.

One adjustment to the suggested snippet: `layer.input_shape` isn't available on Keras 3 layers (it raises `AttributeError`), so the time dimension is read from `layer._build_shapes_dict["sequences_shape"]` instead. Added `test_static_recurrent_with_dynamic_nonrecurrent_input_exports` covering exactly this case.

**Comment by MarcosAsh** [comment]
Fixed Gemini's comments.

**Comment by MarcosAsh** [comment]
I think the numpy fault is just a flakey test. I can commit a blank commit to re-trigger the CI if you wish?

**Review by hertschuh — APPROVED** [review]
Thanks for the fix!
